File: zemiao/123.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        hold1, hold2, sell1, sell2 = -32768,-32768,-32768,-32768
        for p in prices:
            sell2 = max(sell2,hold2+p)
            hold2 = max(hold2,sell1-p)
            sell1 = max(sell1,hold1+p)
            hold1 = max(hold1,-p)
        
        return max(sell1,sell2,0)

File: ztuo/123.java
class Solution {
    public int maxProfit(int[] prices) {
        int buyin1 = Integer.MIN_VALUE;
        int buyin2 = Integer.MIN_VALUE;
        int sell1 = 0;
        int sell2 = 0;
        for (int i : prices) {
            
            sell2 = Math.max(sell2, i + buyin2);//start from the backward sequence
            buyin2 = Math.max(sell1 - i, buyin2); 
            
            sell1 = Math.max(sell1, i + buyin1);
            buyin1 = Math.max(-i, buyin1);                      
        }
        return sell2;       
    }
}